code/modules/actionspeed/_actionspeed_modifier.dm

```diff
@@ -0,0 +1,175 @@
+/*! Actionspeed modification datums.
+
+	How action speed for mobs works
+
+Action speed is now calculated by using modifier datums which are added to mobs. Some of them (nonvariable ones) are globally cached, the variable ones are instanced and changed based on need.
+
+This gives us the ability to have multiple sources of actionspeed, reliabily keep them applied and remove them when they should be
+
+THey can have unique sources and a bunch of extra fancy flags that control behaviour
+
+Previously trying to update action speed was a shot in the dark that usually meant mobs got stuck going faster or slower
+
+Actionspeed modification list is a simple key = datum system. Key will be the datum's ID if it is overridden to not be null, or type if it is not.
+
+DO NOT override datum IDs unless you are going to have multiple types that must overwrite each other. It's more efficient to use types, ID functionality is only kept for cases where dynamic creation of modifiers need to be done.
+
+When update actionspeed is called, the list of items is iterated, according to flags priority and a bunch of conditions
+this spits out a final calculated value which is used as a modifer to last_move + modifier for calculating when a mob
+can next move
+
+*/
+
+/datum/actionspeed_modifier
+	/// Whether or not this is a variable modifier. Variable modifiers can NOT be ever auto-cached. ONLY CHECKED VIA INITIAL(), EFFECTIVELY READ ONLY (and for very good reason)
+	var/variable = FALSE
+
+	/// Unique ID. You can never have different modifications with the same ID. By default, this SHOULD NOT be set. Only set it for cases where you're dynamically making modifiers/need to have two types overwrite each other. If unset, uses path (converted to text) as ID.
+	var/id
+
+	/// Higher ones override lower priorities. This is NOT used for ID, ID must be unique, if it isn't unique the newer one overwrites automatically if overriding.
+	var/priority = 0
+	var/flags = NONE
+
+	/// Multiplicative slowdown
+	var/multiplicative_slowdown = 0
+
+	/// Other modification datums this conflicts with.
+	var/conflicts_with
+
+/datum/actionspeed_modifier/New()
+	. = ..()
+	if(!id)
+		id = "[type]" //We turn the path into a string.
+
+GLOBAL_LIST_EMPTY(actionspeed_modification_cache)
```

Comment:
move this to the top of file idk, don't declare global lists in middle of file

Reply:
done